### PR TITLE
Fix calculating remaining free allowance for SMS

### PR DESCRIPTION
**Pull request by CrystalPea**
The way it was done before, the remainder was incorrect in the billing report and in the org usage query - it was the sms remainder left at the start of the report period, not at the end of that period.

This became apparent when we tried to show sms_remainder on the org usage report, where start date is always the start of the financial year.
We saw that sms sent by services did not reduce their free allowance remainder according to the report. As a result of this, we had to temporarily remove of sms_remainder column from the report, until we fix the bug - it has been fixed now, yay!

I think the bug has snuck in partially because our fixtures for testing this part of the code are quite complex, so it was harder to see that numbers don't add up. I have added comments to the tests to try and make it a bit clearer why the results are as they are.

I also added comments to the code, and renamed some variables, to make it easier to understand, as there are quite a few moving parts in it - subqueries and the like. I also renamed the `fetch_sms_free_allowance_remainder` method to `fetch_sms_free_allowance_remainder_until_date` so it is clearer what it does.

### CODE REVIEW

**Comment by benthorner** [comment]
I started having a look at this. Pea's comment are helpful. Even so, I'm still finding the number of partial / subqueries really hard to follow, since they depend on joins / groupbys that only happen at the last moment.

I printed out the full query SQL as help with reviewing next week:

```
SELECT organisation.name AS organisation_name,
       organisation.id AS organisation_id,
       services.name AS service_name,
       services.id AS service_id,
       anon_1.free_sms_fragment_limit AS anon_1_free_sms_fragment_limit,
       ft_billing.rate AS sms_rate,
       greatest(coalesce(anon_1.sms_remainder, anon_1.free_sms_fragment_limit) - sum(ft_billing.billable_units * ft_billing.rate_multiplier), %(greatest_1)s) AS sms_remainder,
       sum(ft_billing.billable_units * ft_billing.rate_multiplier) AS sms_billable_units,
       greatest(sum(ft_billing.billable_units * ft_billing.rate_multiplier) - coalesce(anon_1.sms_remainder, anon_1.free_sms_fragment_limit), %(greatest_2)s) AS chargeable_billable_sms,
       greatest(sum(ft_billing.billable_units * ft_billing.rate_multiplier) - coalesce(anon_1.sms_remainder, anon_1.free_sms_fragment_limit), %(greatest_2)s) * ft_billing.rate AS sms_cost
FROM services
LEFT OUTER JOIN
  (SELECT annual_billing.service_id AS service_id,
          annual_billing.free_sms_fragment_limit AS free_sms_fragment_limit,
          coalesce(sum(ft_billing.billable_units * ft_billing.rate_multiplier), %(coalesce_1)s) AS billable_units,
          greatest(CAST(annual_billing.free_sms_fragment_limit - coalesce(sum(ft_billing.billable_units * ft_billing.rate_multiplier), %(coalesce_1)s) AS INTEGER), %(greatest_3)s) AS sms_remainder
   FROM annual_billing
   LEFT OUTER JOIN ft_billing ON annual_billing.service_id = ft_billing.service_id
   AND ft_billing.bst_date >= %(bst_date_1)s
   AND ft_billing.bst_date < %(bst_date_2)s
   AND ft_billing.notification_type = %(notification_type_1)s
   WHERE annual_billing.financial_year_start = %(financial_year_start_1)s
   GROUP BY annual_billing.service_id,
            annual_billing.free_sms_fragment_limit) AS anon_1 ON services.id = anon_1.service_id
LEFT OUTER JOIN organisation ON organisation.id = services.organisation_id
JOIN ft_billing ON ft_billing.service_id = services.id
WHERE ft_billing.bst_date >= %(bst_date_3)s
  AND ft_billing.bst_date <= %(bst_date_4)s
  AND ft_billing.notification_type = %(notification_type_2)s
GROUP BY organisation.name,
         organisation.id,
         services.id,
         services.name,
         anon_1.free_sms_fragment_limit,
         anon_1.sms_remainder,
         ft_billing.rate
ORDER BY organisation.name,
         services.name
```

File: app/dao/fact_billing_dao.py [hunk 1]
```diff
@@ -65,31 +65,40 @@ def fetch_sms_free_allowance_remainder(start_date):
 def fetch_sms_billing_for_all_services(start_date, end_date):
 
     # ASSUMPTION: AnnualBilling has been populated for year.
-    free_allowance_remainder = fetch_sms_free_allowance_remainder(start_date).subquery()
+    allowance_left_at_start_date_query = fetch_sms_free_allowance_remainder_until_date(start_date).subquery()
```
Comment (servingUpAces) [inline app/dao/fact_billing_dao.py:68]: How does this fix the bug? We are still passing in the `start_date` rather than `the end_date`

**Review by benthorner — APPROVED** [review]
Nice find and fix @CrystalPea 👏 

I agree the new behaviour is the correct behaviour ✅. We should be returning the up-to-date remainder, not the one at the beginning of the period.

It took me over an hour to understand the query, so I appreciate your effort to make it a bit simpler by clarifying some of the names involved. For the tests, I've suggested how we could be even more helpful to our future selves.

Out-of-scope: something that concerns me about this query is it's coupled to the `rate` in `FactBilling`. If the rate changes during the period, we'll get multiple results for a service, and the data will be...interesting. I couldn't see a test for this scenario. As long as we only change rates at the start of the financial year, I guess it's fine. It is a big assumption, though, so might be worth one of [those big "assumption" comments](https://github.com/alphagov/notifications-api/pull/3395/files#diff-8e5521edad2738f137b117c5cec89497247a508a30bd21065dfe7e8d44aa8dddR67).

File: app/dao/fact_billing_dao.py [hunk 2]
```diff
@@ -65,31 +65,40 @@ def fetch_sms_free_allowance_remainder(start_date):
...
 
     sms_billable_units = func.sum(FactBilling.billable_units * FactBilling.rate_multiplier)
-    sms_remainder = func.coalesce(
-        free_allowance_remainder.c.sms_remainder,
-        free_allowance_remainder.c.free_sms_fragment_limit
+
+    allowance_left_at_start_date = func.coalesce(
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/dao/fact_billing_dao.py:72]: you are right, it's not needed I will get rid of it 👍🏼 .

File: app/dao/fact_billing_dao.py (hunk 1 above)
Comment (CrystalPea) [reply app/dao/fact_billing_dao.py:68]: yeah, sorry, should have split this into commits more, I was renaming things to help myself make sense of what is happening as I was problem solving, and that was a bit of a messy process, hence single commit 😅 

As @benthorner said, `sms_allowance_left` is what we return as `sms_remainder` for the report, while `allowance_left_at_start_date` is still needed because we use it to calculate `chargeable_sms`.

File: tests/app/dao/test_ft_billing_dao.py [hunk 3]
```diff
@@ -592,10 +592,22 @@ def test_fetch_sms_billing_for_all_services_with_remainder(notify_db_session):
 
     results = fetch_sms_billing_for_all_services(datetime(2019, 5, 1), datetime(2019, 5, 31))
     assert len(results) == 3
-    assert results[0] == (org.name, org.id, service.name, service.id, 10, Decimal('0.11'), 8, 3, 0, Decimal('0'))
+
+    # Field names are: organisation_name, organisation_id, service_name, service_id, free_sms_fragment_limit,
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/dao/test_ft_billing_dao.py:596]: Ahh that's a good suggestion 👍🏼 . I will improve the tests and let you know when I'm done.

[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/dao/test_ft_billing_dao.py]: Hey @benthorner I took a stab at changing the tests, they look to be failing because we are comparing a tuple to odict_values(). Do you have an idea for an elegant way to solve this, or should I try a loop of assertions or something? Feel free to add a commit if you find something that works ☺️

[the remark this answers is not in the document]
Comment (benthorner) [reply tests/app/dao/test_ft_billing_dao.py]: @CrystalPea I had a quick play. You could do this:

```
expected_results = [
  { 'organisation_name': org.name, ... },
  ...
]

assert [dict(result) for result in results] == expected_results
```

Basically the same idea, but converting each result to a `dict` first so there aren't any type issues.

Thanks for taking the time to try and refactor. Let me know if this doesn't work and you need a hand.

[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/dao/test_ft_billing_dao.py]: This works, thanks a lot @benthorner ! I think this PR is ready to go now ^_^
